### ISSUE TITLE
Best way to mix online and offline updates

### ISSUE THREAD

**Issue by cajun-rat** [uptane/aktualizr#74]
Offline updates (#8) will allow a single device to get update instructions from two sources:

- The Director
- Someone plugging a USB stick (or whatever) into the device

If only one or the other of these two happens, the behaviour of the system is pretty clear. However we need to consider what will happen in scenarios where the two are interleaved. I assume this to be rare in normal operation, so I think our goals are to introduce this without breaking other things, rather than to support fancy use cases.

I wanted to open this ticket so we can have this "what happens" discussion in the open rather than via email. Here are the considerations I'm aware of so far:

- SotaUptaneClient splits the update into parts like 'Fetch Metadata' 'Download' etc. Those are probably the 'atomic' operations that we need to sequence, so this isn't about locking in SotaUptaneClient, rather whether it OK to fetch the metadata offline, then download /
install online? I think the answer to this question is 'no', but there are more marginal cases to consider.
- The Director currently assumes that once it has handed out a Correlation ID, Aktualizr will always report it back in a manifest with either success / failure. How does this work when the user plugs in a USB stick?
- If the user leaves a USB stick plugged in, does that block online updates?
- It seams reasonable that there are no failure modes of online updates that can indefinitely block an offline update, and vice-versa.
- Whatever the rules are, we need to succinctly define them in order to answer questions like "Is this even meant to work at all?" 
 - A simple implementation would be less work and less maintenance
 - We need clear invariants to avoid this being a source of security holes.

I hope we can work out a proposal for what these rules should be in this ticket!

**Comment by pattivacek** [comment]
Sounds like a good idea! I'll share my first reactions:

> * whether it OK to fetch the metadata offline, then download / install online? I think the answer to this question is 'no', but there are more marginal cases to consider.

I think we should prevent mixing and matching. That sounds like it could open up risky doors and create problems. I'm not sure how best to prevent that, though. I can also see a case for more or less interrupting an online update that's blocking or broken somehow with an offline update that might restore things to a better place.

> * The Director currently assumes that once it has handed out a Correlation ID, Aktualizr will always report it back in a manifest with either success / failure. How does this work when the user plugs in a USB stick?

That I really don't know, but the offline updates don't need a correlation ID, right? What happens if we just leave that out of the equation? Will the Director get particularly upset that the image changed without informing it?

> * If the user leaves a USB stick plugged in, does that block online updates?

Probably. How else do we know how to prioritize or trigger the offline updates? That's the deeper question, right?

> * It seams reasonable that there are no failure modes of online updates that can indefinitely block an offline update, and vice-versa.

That seems like a good goal and might even be a given.

**Comment by tkfu** [comment]
The offline updates definitely don't need (or have) a correlation ID. As for director behaviour, I think we should get @simao in this thread (I think github won't let me actually ping him via @ until he joins of his own accord). I think it's technically device registry that keeps track of device state, but state changes are triggered by events emitted by director. Here's the state diagram as I understand it:

```mermaid
stateDiagram-v2
  NotSeen --> UpToDate: device sends manifest
  UpToDate --> Outdated: update is assigned
  Outdated --> UpdatePending: device downloads director metadata w/ correlationId
  Outdated --> UpToDate: assigned update is cancelled
  UpdatePending --> UpToDate: device reports success w/ correlationId
  UpdatePending --> Error: device reports failure w/ correlationId
  Error --> Outdated: update is assigned
```
(by the way: isn't it cool that github added support for inline mermaid diagrams in markdown??)

Note that both transitions out of UpdatePending can only happen when we get a manifest from aktualizr containing an InstallationReport with correlationId in its custom metadata. (Director throws an error if you attempt to assign an update to a device that is currently Outdated or UpdatePending.)

On principle, I think that the decision about whether to interrupt an in-progress update with an update of another type should be left to the user of libaktualizr: the calling program should make decisions about the circumstances under which an update should be preempted. But that leaves us to decide how to construct the manifest in the case where an online update is interrupted by an offline update. I feel like in an ideal world, we would include an InstallationReport for the successful install of an offline update in the next manifest sent (n.b. this isn't something that's currently implemented), probably using the name/version of the offline update installed in lieu of the correlationId. But we also should communicate that the previously-directed online update, if any, was aborted. One way to do that would be to start letting aktualizr send multiple InstallationReports in a single manifest. Of course, director would have to learn to handle that.

**Comment by simao** [comment]
Yes, we are not sending any correlation ids in the lockboxes, I don't think that is needed, and we also don't have it at the time the user creates the lock box.

Regarding the device reporting to director something that was installed offline, director will still process the manifest.

If the device has some assignment in director already, an online update, director will still process the manifest and just update the currently installed ecu target, put keep the device queue intact. If the offline installed update somehow matches the update in the queue (ecuId, filename, checksum), and that assignment in director has a correlation id, then the normal execution for online updates will be followed. If the offline update does not match the online update, then the processing will be the same as when the queue is empty, and the device queue will remain as before.

If the queue for the device is empty, director will just update the current known installed ecu target to the new target reported by device. If the manifest reports an error, this will be published in the bus, and device registry will update the status to error, otherwise the status will be moved to Up To Date.

**Comment by simao** [comment]
In the case of offline updates, the device will not have a correlation id, but it could build one like `urn:here-ota:offline-update:<some-id>` (All others are like `urn:here-ota:auto-update:<uuid>`, `urn:here-ota:mtu:<uuid>` etc. Though this needs some backend changes to handle correctly.
